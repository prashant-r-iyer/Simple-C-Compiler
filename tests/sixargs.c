void printall(long a, long b, long c, long d, long e, long f) {
    printf("%d\n", a);
    printf("%d\n", b);
    printf("%d\n", c);
    printf("%d\n", d);
    printf("%d\n", e);
    printf("%d\n", f);
}

void main() {
    printall(1, 2, 3, 4, 5, 6);
}