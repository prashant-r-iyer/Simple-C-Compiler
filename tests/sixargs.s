	.text

.globl printall
printall:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9
	movq %rdi, -8(%rbp)
	movq %rsi, -16(%rbp)
	movq %rdx, -24(%rbp)
	movq %rcx, -32(%rbp)
	movq %r8, -40(%rbp)
	movq %r9, -48(%rbp)
	movq $string0, %rbx
	movq -8(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	movq $string1, %rbx
	movq -16(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	movq $string2, %rbx
	movq -24(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	movq $string3, %rbx
	movq -32(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	movq $string4, %rbx
	movq -40(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	movq $string5, %rbx
	movq -48(%rbp), %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

	.text

.globl main
main:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9

	# push 1
	movq $1, %rbx

	# push 2
	movq $2, %r10

	# push 3
	movq $3, %r13

	# push 4
	movq $4, %r14

	# push 5
	movq $5, %r15

	# push 6
	movq $6, %r11
	# func=printall nargs=6

	# Move values from reg stack to reg args
	movq %r11, %r9
	movq %r15, %r8
	movq %r14, %rcx
	movq %r13, %rdx
	movq %r10, %rsi
	movq %rbx, %rdi
	call printall
	movq %rax, %rbx

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

string0:
	.string "%d\n"

string1:
	.string "%d\n"

string2:
	.string "%d\n"

string3:
	.string "%d\n"

string4:
	.string "%d\n"

string5:
	.string "%d\n"

