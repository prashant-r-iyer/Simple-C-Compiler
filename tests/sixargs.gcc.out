sixargs.c: In function ‘printall’:
sixargs.c:2:5: warning: implicit declaration of function ‘printf’ [-Wimplicit-function-declaration]
    2 |     printf("%d\n", a);
      |     ^~~~~~
sixargs.c:1:1: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
  +++ |+#include <stdio.h>
    1 | void printall(long a, long b, long c, long d, long e, long f) {
sixargs.c:2:5: warning: incompatible implicit declaration of built-in function ‘printf’ [-Wbuiltin-declaration-mismatch]
    2 |     printf("%d\n", a);
      |     ^~~~~~
sixargs.c:2:5: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
sixargs.c:2:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    2 |     printf("%d\n", a);
      |             ~^     ~
      |              |     |
      |              int   long int
      |             %ld
sixargs.c:3:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    3 |     printf("%d\n", b);
      |             ~^     ~
      |              |     |
      |              int   long int
      |             %ld
sixargs.c:4:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    4 |     printf("%d\n", c);
      |             ~^     ~
      |              |     |
      |              int   long int
      |             %ld
sixargs.c:5:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    5 |     printf("%d\n", d);
      |             ~^     ~
      |              |     |
      |              int   long int
      |             %ld
sixargs.c:6:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    6 |     printf("%d\n", e);
      |             ~^     ~
      |              |     |
      |              int   long int
      |             %ld
sixargs.c:7:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    7 |     printf("%d\n", f);
      |             ~^     ~
      |              |     |
      |              int   long int
      |             %ld
